src/utils/context_free_grammar_generator.cpp

```diff
@@ -125,36 +116,66 @@ context_free_grammar_generator::context_free_grammar_generator(const std::map<st
 			}
 		}
 	}
-	initialized_ = true;
 }
 
-std::string context_free_grammar_generator::print_nonterminal(const std::string& name, uint32_t* seed, short seed_pos) const {
-	std::string result;
-	std::map<std::string, nonterminal>::const_iterator found = nonterminals_.find(name);
-	if (found == nonterminals_.end()) {
-		lg::wml_error() << "[context_free_grammar_generator] Warning: needed nonterminal " << name << " not defined";
-		return "!" + name;
+std::string context_free_grammar_generator::print_nonterminal(const std::string& name, const std::map<std::string,std::string>* variables, uint32_t* seed, short seed_pos) const {
+	if (name.length() >= 2 && name[0] == '$') {
+		std::string variable = name.substr(1);
+		if (variables == nullptr) {
+			goto warn_missing_variable;
```

Comment:
It looks like you could eliminate these two goto statements by simply inverting your conditionals.